docs/plugin_developers_guide/01_Fledge_plugins.rst

```diff
@@ -41,11 +41,136 @@ Plugins
-- **Notification Delivery Plugins** - These plugins are used by the optional notification service to deliver a notification to a system when a notification rule has triggered. These plugins allow the mechanisms to deliver notifications to be extended.
+- **Filter Plugins** - Filter plugins are used to modify data as it flows through Fledge. One or more filter plugins may compose a pipeline which modifies, either, data flowing out from the South ingress service into the Fledge Storage system, or out from Fledge storage into the North egress tasks.
+- **Notification Rule Plugins** - Notification plugins evaluate data in parallel with their entry into the Fledge Storage, before it enters the northbound Filter pipeline(s). If the notification "fires", notification delivery is executed. Rules can fire for triggering or clearing of the rule condition.
+  Notification rule plugins are applied to the notification service, the notification service is an optional service and must be installed before the notification plugins can be used.
+- **Notification Delivery Plugins** - These plugins deliver a notification event; events, trigger "reason", and accompanying "message" are specified, and delivered to a given [notification delivery type](#notification-delivery-types). Notification delivery plugins require installation of the optional notification service.
+- **North Plugins** - These plugins take data moving through Fledge and exports the data to the outside world. Data may originate from South plugins, from the Fledge system, or be added by filters.
+
+Message contents
+================
+Information flows through the system in "messages". Messages have a few "tags" which have well defined values. Messages may be extended to contain additional tags appropriate to particular plugins and data flows.
+
+Most messages include:
+
++------------+------------------+-------------------------------------------------------------------------------------------------------+
+| Tag        | Type             | Description                                                                                           |
++============+==================+=======================================================================================================+
+| asset      | string           | Name under which this stream of data are placeed in Storage                                           |
++------------+------------------+-------------------------------------------------------------------------------------------------------+
+| timestamp  | string           | Fledge timestamp when this asset was generated                                                        |
++------------+------------------+-------------------------------------------------------------------------------------------------------+
+| key        | string           | unique ID                                                                                             |
```

Comment:
This is added by the storage when a reading is stored. Readings only have an ID when you view them via the REST API or in north tasks.